packages/pipeline/src/mocks/IntersectionObserver.ts

```diff
@@ -0,0 +1,91 @@
+import { act } from 'react-dom/test-utils';
+import { IntersectionOptions } from 'react-intersection-observer';
+
+const observerMap = new Map();
+const instanceMap = new Map();
+
+function IntersectionObserver(
```

Comment:
Curious why we wouldn't use a [polyfill](https://github.com/w3c/IntersectionObserver/tree/master/polyfill) for this?

Reply:
I'm sure we can use a polyfill, I am not that familiar with how Jest works / is setup to include a polyfill. I would appreciate some guidance in how to do this.

Reply:
We have `scripts/jestSetup.js` which uses the [setupFiles](https://jestjs.io/docs/en/configuration#setupfiles-array) configuration of Jest for configuring the environment of tests.

Reply:
This is related to the way in which I am testing the Stage component. I use this as a mock to test the functionality for determining the visibility of the `Stage` component relative to its parent.

<details open>
<summary>I had attempted to test this in an alternative manner:</summary>
<p>

```ts
describe('when it is NOT intersecting its parent', () => {
    test('should set the "data-stage-visible" attribute to "false"', async () => {
      const { element } = renderStage();

      // Set the starting sizes of the element under test and the parentElement container
      element.setAttribute('style', 'width: 50px; margin-left: -100px;');
      parentElement.setAttribute('style', 'width: 100px;');

      // Observe changes to the element
      const promise = waitForDomChange({
        container: element,
        mutationObserverOptions: {
          attributes: true,
          childList: false,
          subtree: false,
        },
      });

      // Trigger an intersection change on the element and await the result
      element.setAttribute('style', 'width: 100px; margin-left: 0px;');
      await promise;

      // Make assertion on element visibility attribute
      expect(element.getAttribute('data-stage-visible')).toBe('true');
    });
  });

  describe('when it is intersecting its parent', () => {
    test('should set the "data-stage-visible" attribute to "false"', async () => {
      const { element } = renderStage();

      // Set the starting sizes of the element under test and the parentElement container
      element.setAttribute('style', 'width: 50px; margin-left: 0px;');
      parentElement.setAttribute('style', 'width: 100px;');

      // Observe changes to the element
      const promise = waitForDomChange({
        container: element,
        mutationObserverOptions: {
          attributes: true,
          childList: false,
          subtree: false,
        },
      });

      // Trigger an intersection change on the element and await the result
      element.setAttribute('style', 'width: 50px; margin-left: -100px;');
      await promise;

      // Make assertion on element visibility attribute
      expect(element.getAttribute('data-stage-visible')).toBe('false');
    });
  });
```

</p>
</details>

But I get a test failure no matter what I do, so something is clearly not write with the way in which this test is written:

```
packages/pipeline/Stage › when it is NOT intersecting its parent › should set the "data-stage-visible" attribute to "false"

Expected: "true"
Received: "false"
```

If this is a blocker for merging, then I would really appreciate some help as I am all out of ideas how to best solve this.

Reply:
Adding some context on @matt-d-rat behalf here^:

Jest is using jsdom to configure a browser-like environment in Node, however it [doesn't actually support layout](https://github.com/jsdom/jsdom/blob/master/README.md#unimplemented-parts-of-the-web-platform) or methods like `getBoundingClientRects()` which the polyfill you linked uses heavily.

As a result, even with the polyfill, code that tests behaviour that relies on the IntersectionObserver just won't work.

The alternatives that remain would be to either:
* run the tests using something like jest-puppeteer
* mock out `getBoundingClientRects()` and any other layout/size methods used by the IntersectionObserver
* mock out the IntersectionObserver, which Matt did here

Reply:
Thanks @scott I was having trouble wording 🤣

Reply:
Ah, didn't know that the polyfill relied on `getBoundingClientRects`. That makes complete sense.

I'm fine with either mocking out `getBoundingClientRects` or using the mock that @matt-d-rat made, up to ya'll which you prefer.

If we go with the current mock can we move it somewhere more general so we can re-use it across other packages if needed? We have `scripts/jestSetup.js` which has one mock created, so ideally we can build on top of that (make a folder for this file, inline the file, up to you).

Reply:
I'll stick with the mock I have implemented already, and attempt to move it to a re-usable location.

Reply:
I attempted to move the file to a re-usable location, but when I attempt to build the project I get the following error:

```bash
$ tsc --build tsconfig.json
packages/pipeline/src/Stage.spec.tsx:2:36 - error TS6059: File '/Users/matt/Documents/projects/mongodb/leafygreen-ui/__mocks__/IntersectionObserver.ts' is not under 'rootDir' '/Users/matt/Documents/projects/mongodb/leafygreen-ui/packages/pipeline'. 'rootDir' is expected to contain all source files.

2 import { mockIsIntersecting } from '../../../__mocks__/IntersectionObserver';
```

I also tried a variety of jest config options as well (`moduleDirectories`,  `modulePaths`), which complained that the file could not be found when importing. If someone with more jest experience than me could take a look at this then I would appreciate it.

The relevant code that needs moving is located under `packages/pipeline/mocks`

Reply:
Changed my mind - let's worry about moving this file when we have a need for it. I keep running out of time to look into it.